### PR TITLE
Manage Python project

### LINKED ISSUE #12: Python package structure
##  Advanced

Here is a list of pre-commit checks that I typically use for my project:

- Linters: [flake8](https://github.com/pycqa/flake8), [pylint](https://github.com/PyCQA/pylint). My choice: [flake8]. Additional flake8-plugins:
    - [flake8-bugbear](https://pypi.org/project/flake8-bugbear/) for detecting bugs and potential design problems
    - [flake8-builtins](https://pypi.org/project/flake8-builtins/) for verifying that builtins are not used as variable names
    - [flake8-comprehensions](https://pypi.org/project/flake8-comprehensions/) for writing better and consistent comprehensions
    - [flake8-debugger](https://pypi.org/project/flake8-debugger/) for checking that there are no forgotten [breakpoints](https://docs.python.org/3/library/functions.html#breakpoint)
    - [flake8-eradicate](https://pypi.org/project/flake8-eradicate/) to avoid “dead” or commented code
    - [flake8-logging-format](https://github.com/globality-corp/flake8-logging-format) for consistent logging
    - [pep8-naming](https://pypi.org/project/pep8-naming/) for verifying that PEP8 naming conventions are followed
    - [tryceratops](https://pypi.org/project/tryceratops/) to avoid exception anti-patterns
- Type checking: [mypy](https://github.com/python/mypy). My choice: [[mypy](https://github.com/python/mypy)]
    - [types-click-spinner](https://github.com/click-contrib/click-spinner)
    - [types-requests, types-tabulate, types-toml](https://github.com/python/typeshed)
- Auto-formatters:
    - [isort](https://pycqa.github.io/isort/) for sorting imports
    - [autoflake](https://github.com/PyCQA/autoflake) for removing unused imports
    - [black](https://github.com/psf/black) for formatting the rest of the code.
- Python auto-update:  [pyupgrade](https://github.com/asottile/pyupgrade) tool to automatically upgrade syntax for newer versions of the language.

### LINKED ISSUE #124: Testing

### PR DESCRIPTION
**Pull request by yakutovicha**
fixes #12
fixes #124 

ToDo:

- [x] Fix `toc.py`

### CODE REVIEW

File: manage_python_project.ipynb [hunk 1]
```diff
@@ -0,0 +1,1760 @@
...
+    "However, as usual, it is common to install `pytest-cov` as part of the development dependencies, see [above](#Optional-dependencies).\n",
+    "\n",
+    "\n",
+    "To achieve 100% test coverage, we need to test both branches:\n",
+    "    \n",
+    "    ```python\n",
+    "    def test_do_something():\n",
+    "        assert do_something(True) == 1\n",
+    "        assert do_something(False) == 2\n",
+    "    ```\n",
```
Comment (edoardob90) [inline manage_python_project.ipynb:728]:
```suggestion
    "\n",
    "```python\n",
    "def test_do_something():\n",
    "    assert do_something(True) == 1\n",
    "    assert do_something(False) == 2\n",
    "```\n",
```

File: manage_python_project.ipynb [hunk 2]
```diff
@@ -0,0 +1,1760 @@
...
+    "    assert divide(1, 2) == 0.5\n",
+    "```\n",
+    "\n",
+    "\n",
+    "The `assert` statement checks that the condition is `True`.\n",
+    "If the condition is `False`, the `assert` statement raises an `AssertionError` exception and the test fails.\n",
+    "\n",
+    "\n",
+    "<div class=\"alert alert-block alert-info\">\n",
+    "Test is considered successful if no exceptions are raised during its execution.\n",
```
Comment (edoardob90) [inline manage_python_project.ipynb:634]:
```suggestion
    "A test is considered successful if <strong>no exceptions</strong> are raised during its execution.\n",
```

File: manage_python_project.ipynb [hunk 3]
```diff
@@ -0,0 +1,1760 @@
...
+    "- Tests: Fix `StructureData` test breaking for recent `pymatgen` versions [[093037d48]](https://github.com/aiidateam/aiida-core/commit/093037d48a2d92cbb6f068c1111fe1564a4500c0)\n",
+    "- Trigger Docker image build when pushing to `support/*` branch [[5cf3d1d75]](https://github.com/aiidateam/aiida-core/commit/5cf3d1d75e8d22d6a3f0909c84aa63cc228bcf4b)\n",
+    "- Use `aiida-core-base` image from `ghcr.io` [[0e5b1c747]](https://github.com/aiidateam/aiida-core/commit/0e5b1c7473030dd5b5027ea4eb0a658db9174091)\n",
+    "- Loosen trigger conditions for Docker build CI workflow [[22e8a8069]](https://github.com/aiidateam/aiida-core/commit/22e8a80690747b792b70f96a0e332906f0e65e97)\n",
+    "- Follow-up docker build runner macOS-ARM64 [[1bd9bf03d]](https://github.com/aiidateam/aiida-core/commit/1bd9bf03d19dda4c462728fb87cf4712b74c5f39)\n",
+    "- Upload artifact by PR from forks for docker workflow [[afc2dad8a]](https://github.com/aiidateam/aiida-core/commit/afc2dad8a68e280f01e89fcb5b13e7a60c2fd072)\n",
+    "- Update the image name for docker image [[17507b410]](https://github.com/aiidateam/aiida-core/commit/17507b4108b5dd1cd6e074b08e0bc2535bf0a164)\n",
+    "```"
```
Comment (edoardob90) [inline manage_python_project.ipynb:1459]:
```suggestion
    "```\n",
    "\n",
    "There are also some tools that you can use to automatically build a changelog of a git repository. If you are interested, have a look at [git-cliff](https://github.com/orhun/git-cliff)."
```

File: manage_python_project.ipynb [hunk 4]
```diff
@@ -0,0 +1,1760 @@
...
+   "metadata": {},
+   "source": [
+    "# Automate things\n",
+    "\n",
+    "Most of the things that we have discussed in this tutorial can (and should) be automated.\n",
+    "In this section we will discuss how to automate the following things:\n",
+    "\n",
+    "- Testing.\n",
+    "- Version update.\n",
+    "- Releasing."
```
Comment (edoardob90) [inline manage_python_project.ipynb:1501]:
```suggestion
    "- Testing\n",
    "- Version update\n",
    "- Releasing"
```

File: manage_python_project.ipynb [hunk 5]
```diff
@@ -0,0 +1,1760 @@
...
+  {
+   "cell_type": "markdown",
+   "id": "ca42772a",
+   "metadata": {},
+   "source": [
+    "## Releasing\n",
+    "\n",
+    "To automate the release process we will use [GitHub Actions](https://docs.github.com/en/actions).\n",
+    "We already did big part of the work in the section above, when talking about version update.\n",
+    "The only missing part is to initial the GitHub machinery to automate the release process both on GitHub and on PyPI.\n",
```
Comment (edoardob90) [inline manage_python_project.ipynb:1641]:
```suggestion
    "The only missing part is to set up GitHub's machinery to automate the release process on both GitHub and PyPI.\n",
```

**Review by edoardob90 — CHANGES_REQUESTED** [review]
Read through it all, and it's great! 🙌🏻 

I suggested only a few minor changes that can be approved on-the-fly.

File: manage_python_project.ipynb (hunk 2 above)
Comment (yakutovicha) [reply manage_python_project.ipynb:634]: done

File: manage_python_project.ipynb (hunk 3 above)
Comment (yakutovicha) [reply manage_python_project.ipynb:1459]: done

File: manage_python_project.ipynb (hunk 4 above)
Comment (yakutovicha) [reply manage_python_project.ipynb:1501]: done

File: manage_python_project.ipynb (hunk 5 above)
Comment (yakutovicha) [reply manage_python_project.ipynb:1641]: done.

**Review by edoardob90 — APPROVED** [review]
Let's go 🚀

**Pull request merged**